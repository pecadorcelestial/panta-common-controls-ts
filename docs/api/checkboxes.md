# CheckBox

### Properties

| Name    | Type | Required | Comments |
| --------|------|----------|----------|
| checked | bool | false    | Sets the initial state of the **checkbox**. |

### Functions

| Name     | Parameters | Type | Description |
|----------|------------|------|-------------|
| onChange | checked    | bool | Returns the state of the **checkbox** when it changes. |

### Methods

| Name     | Parameters | Type | Description |
|----------|------------|------|-------------|
| getValue | N/A        | N/A  | Returns the state of the **checkbox**. |
| setValue | value      | bool | Changes the value of the **checkbox**. |

## Usage

```javascript
import React, { Component } from 'react';
import { CheckBox } from 'panda-common-controls';

class BasicForm extends Component {
    render() {
        //Properties.
        const props = {
            //Required.
            checked: false,
            //Functions.
            onChange: (checked) => {console.log('[panda-common-controls][test][onChange] Valor: ', checked);}
        };
        //Result.
        return(
            <div>
                <CheckBox {...props}>Â¿Should this be checked?</CheckBox>
            <div>
        );
    }
}
```
