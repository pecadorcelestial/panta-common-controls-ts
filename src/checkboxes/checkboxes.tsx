//Componentes generales.
import React, { Component, HTMLProps } from 'react';
import styled, { StyledComponent } from 'styled-components';

const Layout = styled.div`
    box-sizing: border-box;
	display: inline-block;
	height: 30px;
	margin: 0px;
	padding: 0px;
    width: auto;
`;

interface ICheckMarkProps {
    checked: boolean;
};
const CheckMark: StyledComponent<'div', any, ICheckMarkProps & HTMLProps<HTMLDivElement>, never> = styled.div`
    background-color: ${(props: ICheckMarkProps & HTMLProps<HTMLDivElement>) => props.checked ? `#1476FB`: `#BFBFBF`};
    border: none;
    border-radius: 5px;
    box-sizing: border-box;
    cursor: pointer;
    display: inline-block;
    float: left;
    height: 20px;
    margin: 5px 0px 0px 0px;
    width: 20px;
    
    ${Layout}:hover & {
        background-color: ${(props: ICheckMarkProps & HTMLProps<HTMLDivElement>) => props.checked ? `#11AAFF`: `#A7A7A7`};
    }

    ${(props: ICheckMarkProps & HTMLProps<HTMLDivElement>) => props.checked ? `
        &:after {
            border: solid white;
            border-width: 0px 3px 3px 0px;
            content: "";
            display: block;
            height: 10px;
            margin: 2px 0px 0px 6px;
            width: 5px;

            transform: rotate(45deg);
        }` : ``}
`;

const Title: StyledComponent<'label', any, HTMLProps<HTMLLabelElement>, never> = styled.label`
    box-sizing: border-box;
    color: #242424;
    cursor: pointer;
    display: inline-block;
    float: left;
	height: 30px;
	font-family: "Open Sans", sans-serif;
	font-size: 16px;
	font-weight: normal;
	font-style: normal;
	font-stretch: normal;
	letter-spacing: normal;
	margin: 0px;
	padding: 4px 0px 0px 5px;
	text-align: left;
	width: auto;
    white-space: nowrap;
`;

interface ICheckBoxProps extends ICheckMarkProps {
    onChange?: Function;
};
interface ICheckBoxState {
    checked: boolean;
};
export class CheckBox extends Component<ICheckBoxProps, ICheckBoxState> {
    //*** CONSTRUCTOR ***
    constructor(props: ICheckBoxProps) {
        super(props);
        this.state = {
            checked: this.props.checked
        };
    }
    //*** MÃ‰TODOS ***
    getValue = (): boolean => {
        return this.state.checked;
    }
    setValue = (value: boolean): void => {
        this.setState({ checked: value });
    }
    //*** RESULTADO ***
    render() {
        return(
            <Layout onClick={() => { let checked = !this.state.checked; this.setState({ checked }); if(this.props.onChange) this.props.onChange(checked); }}>
                <CheckMark checked={this.state.checked}/>
                <Title checked={this.state.checked}>{this.props.children}</Title>
            </Layout>
        );
    }
}