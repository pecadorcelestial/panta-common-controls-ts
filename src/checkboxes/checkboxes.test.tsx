
//RRRR  EEEEE  AAA   CCCC TTTTT
//R   R E     A   A C       T
//RRRR  EEE   AAAAA C       T
//R   R E     A   A C       T
//R   R EEEEE A   A  CCCC   T

import React from 'react';
import renderer from 'react-test-renderer';

//EEEEE N   N ZZZZZ Y   Y M   M EEEEE
//E     NN  N    Z   Y Y  MM MM E
//EEE   N N N   Z     Y   M M M EEE
//E     N  NN  Z      Y   M   M E
//EEEEE N   N ZZZZZ  YYY  M   M EEEEE

import * as enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

enzyme.configure({ adapter: new Adapter() });

//JJJJJ EEEEE  SSSS TTTTT
//    J E     S       T
//    J EEE    SSS    T
//J   J E         S   T
// JJJ  EEEEE SSSS    T

//NOTA: Éste módulo ayuda a que los errores en componentes estilizados se vean de manera correcta:
//https://medium.com/styled-components/effective-testing-for-styled-components-67982eb7d42b
import 'jest-enzyme';
import 'jest-styled-components';

// CCCC  OOO  M   M PPPP   OOO  N   N EEEEE N   N TTTTT EEEEE  SSSS
//C     O   O MM MM P   P O   O NN  N E     NN  N   T   E     S
//C     O   O M M M PPPP  O   O N N N EEE   N N N   T   EEE    SSS
//C     O   O M   M P     O   O N  NN E     N  NN   T   E         S
// CCCC  OOO  M   M P      OOO  N   N EEEEE N   N   T   EEEEE SSSS

import { CheckBox } from './checkboxes';

//Snapshots.
describe('[SNAPSHOT][Componentes][Common][CheckBox] - Checkbox normal.', () => {
	//UNCHECKED.
	it('Debe pintar el componente con los siguientes modificadores: (Checked: false).', () => {
		//Funciones 'dummy'.
		const handleOnChange = event => {
			//Nada que hacer aquí.
        };
		//Propiedades.		
        const checkboxProps = {
            //Opcionales.
            disabled: false,
            checked: false,
            //Funciones.
            onChange: handleOnChange
        };
        //Se crea el componente.
        
        //U   U N   N  CCCC H   H EEEEE  CCCC K   K EEEEE DDDD
        //U   U NN  N C     H   H E     C     K  K  E     D   D
        //U   U N N N C     HHHHH EEE   C     KKK   EEE   D   D
        //U   U N  NN C     H   H E     C     K  K  E     D   D
        // UUU  N   N  CCCC H   H EEEEE  CCCC K   K EEEEE DDDD

        //Snapshot.
        const component = renderer.create(<CheckBox {...checkboxProps}/>).toJSON();
		expect(component).toMatchSnapshot();
    });
    //CHECKED.
	it('Debe pintar el componente con los siguientes modificadores: (Checked: false).', () => {
		//Funciones 'dummy'.
		const handleOnChange = event => {
			//Nada que hacer aquí.
        };
		//Propiedades.		
        const checkboxProps = {
            //Opcionales.
            disabled: false,
            checked: true,
            //Funciones.
            onChange: handleOnChange
        };
        //Se crea el componente.
        
        // CCCC H   H EEEEE  CCCC K   K EEEEE DDDD
        //C     H   H E     C     K  K  E     D   D
        //C     HHHHH EEE   C     KKK   EEE   D   D
        //C     H   H E     C     K  K  E     D   D
        // CCCC H   H EEEEE  CCCC K   K EEEEE DDDD

        //Snapshot.
        const component = renderer.create(<CheckBox {...checkboxProps}/>).toJSON();
		expect(component).toMatchSnapshot();
    });
});

//Flujo #1 (2 clics).
describe('[FLUJO][Componentes][Common][CheckBox] - Valida el cambio de estado al dar clic sobre el componente.', () => {
    //Funciones 'dummy'.
    const handleOnChange = jest.fn(event => {
		//Nada que hacer aquí.
	});
    //Propiedades.		
    const checkboxProps = {
        //Opcionales.
        checked: false,
        disabled: false,
        //Funciones.
        onChange: handleOnChange
    };
	//Se crea el componente.
	//NOTA: Al utilizar la función 'mount' se detona las siguientes funciones: constructor, componentDidMount y render.
    const component = enzyme.mount(<CheckBox {...checkboxProps}/>);

	// CCCC  OOO  M   M PPPP   OOO  N   N EEEEE N   N TTTTT DDDD  IIIII DDDD  M   M  OOO  U   U N   N TTTTT
	//C     O   O MM MM P   P O   O NN  N E     NN  N   T   D   D   I   D   D MM MM O   O U   U NN  N   T
	//C     O   O M M M PPPP  O   O N N N EEE   N N N   T   D   D   I   D   D M M M O   O U   U N N N   T
	//C     O   O M   M P     O   O N  NN E     N  NN   T   D   D   I   D   D M   M O   O U   U N  NN   T
	// CCCC  OOO  M   M P      OOO  N   N EEEEE N   N   T   DDDD  IIIII DDDD  M   M  OOO   UUU  N   N   T

	it('Se verifica que la información inicial esté correcta.', () => {
		//Expectativas.
        expect(component.state().checked).toBe(false);
	});

	// OOO  N   N  CCCC L     IIIII  CCCC K   K
	//O   O NN  N C     L       I   C     K  K
	//O   O N N N C     L       I   C     KKK
	//O   O N  NN C     L       I   C     K  K
	// OOO  N   N  CCCC LLLLL IIIII  CCCC K   K

	it('Debe mandar llamar la función "onChange" que se pasa como propiedad, y cambiar el valor "checked" a true.', () => {
		//Simulación.
		component.simulate('click');
		//Expectativa.
        expect(handleOnChange).toHaveBeenCalled();
        expect(component.state().checked).toBe(true);
	});
	it('Debe mandar llamar la función "onChange" que se pasa como propiedad, y cambiar el valor "checked" a false.', () => {
		//Simulación.
		component.simulate('click');
		//Expectativa.
        expect(handleOnChange).toHaveBeenCalled();
        expect(component.state().checked).toBe(false);
    });
});

//Flujo #2 (2 clics con referencia nula).
describe('[FLUJO][Componentes][Common][CheckBox] - Valida el cambio de estado al dar clic sobre el componente.', () => {
    //Funciones 'dummy'.
    const handleOnChange = jest.fn(event => {
		//Nada que hacer aquí.
	});
    //Propiedades.		
    const checkboxProps = {
        //Opcionales.
        checked: false,
        disabled: false
    };
	//Se crea el componente.
	//NOTA: Al utilizar la función 'mount' se detona las siguientes funciones: constructor, componentDidMount y render.
    const component = enzyme.mount(<CheckBox {...checkboxProps}/>);

	// CCCC  OOO  M   M PPPP   OOO  N   N EEEEE N   N TTTTT DDDD  IIIII DDDD  M   M  OOO  U   U N   N TTTTT
	//C     O   O MM MM P   P O   O NN  N E     NN  N   T   D   D   I   D   D MM MM O   O U   U NN  N   T
	//C     O   O M M M PPPP  O   O N N N EEE   N N N   T   D   D   I   D   D M M M O   O U   U N N N   T
	//C     O   O M   M P     O   O N  NN E     N  NN   T   D   D   I   D   D M   M O   O U   U N  NN   T
	// CCCC  OOO  M   M P      OOO  N   N EEEEE N   N   T   DDDD  IIIII DDDD  M   M  OOO   UUU  N   N   T

	it('Se verifica que la información inicial esté correcta.', () => {
		//Expectativas.
        expect(component.state().checked).toBe(false);
	});

	// OOO  N   N  CCCC L     IIIII  CCCC K   K
	//O   O NN  N C     L       I   C     K  K
	//O   O N N N C     L       I   C     KKK
	//O   O N  NN C     L       I   C     K  K
	// OOO  N   N  CCCC LLLLL IIIII  CCCC K   K

	it('No debe mandar llamar la función "onChange" que se pasa como propiedad, y cambiar el valor "checked" a true.', () => {
		//Simulación.
		component.simulate('click');
		//Expectativa.
        expect(handleOnChange).not.toHaveBeenCalled();
        expect(component.state().checked).toBe(true);
	});
	it('No debe mandar llamar la función "onChange" que se pasa como propiedad, y cambiar el valor "checked" a false.', () => {
		//Simulación.
		component.simulate('click');
		//Expectativa.
        expect(handleOnChange).not.toHaveBeenCalled();
        expect(component.state().checked).toBe(false);
    });
});

//Métodos.
describe("[MÉTODOS][Componentes][Common][CheckBox] - Valida las llamadas a los métodos dentro del componente.", () => {
	//Propiedades.		
    const checkboxProps = {
        //Opcionales.
        checked: false,
        disabled: false
    };
	//Se crea el componente.
	//NOTA [1]: Al utilizar la función "mount" se detona la función "componentDidMount".
	//NOTA [2]: Además, con "mount" se puede utilizar el método "attachTo" para utilizar "document".
	const component = enzyme.mount(<CheckBox {...checkboxProps}/>);
	
	// CCCC  OOO  M   M PPPP   OOO  N   N EEEEE N   N TTTTT DDDD  IIIII DDDD  M   M  OOO  U   U N   N TTTTT
	//C     O   O MM MM P   P O   O NN  N E     NN  N   T   D   D   I   D   D MM MM O   O U   U NN  N   T
	//C     O   O M M M PPPP  O   O N N N EEE   N N N   T   D   D   I   D   D M M M O   O U   U N N N   T
	//C     O   O M   M P     O   O N  NN E     N  NN   T   D   D   I   D   D M   M O   O U   U N  NN   T
	// CCCC  OOO  M   M P      OOO  N   N EEEEE N   N   T   DDDD  IIIII DDDD  M   M  OOO   UUU  N   N   T

	//La única propiedad que se revisa en la función 'componentWillReceiveProps' es el valor por defecto.
	it('Se verifica que la información inicial esté correcta.', () => {
		//Expectativas.
		expect(component.state().checked).toBe(false);
	});

    //V   V  AAA  L      OOO  RRRR
    //V   V A   A L     O   O R   R
    //V   V AAAAA L     O   O RRRR
    // V V  A   A L     O   O R   R
    //  V   A   A LLLLL  OOO  R   R

	it('Debe asignar el valor enviado en el control.', () => {
		//Simulación.
		component.instance().setValue(true);
		//Expectativa.
		expect(component.state().checked).toBe(true);
    });
    it('Debe obtener el valor asignado al control.', () => {
        //Simulación.
        let value = component.instance().getValue();
        //Expectativa.
        expect(value).toBe(true);
    });
});
