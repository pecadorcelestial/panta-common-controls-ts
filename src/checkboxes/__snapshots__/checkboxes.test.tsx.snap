// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[SNAPSHOT][Componentes][Common][CheckBox] - Checkbox normal. Debe pintar el componente con los siguientes modificadores: (Checked: false). 1`] = `
.c1 {
  box-sizing: border-box;
  display: inline-block;
  height: 30px;
  margin: 0px;
  padding: 0px;
  width: auto;
}

.c2 {
  background-color: #BFBFBF;
  border: none;
  border-radius: 5px;
  box-sizing: border-box;
  cursor: pointer;
  display: inline-block;
  float: left;
  height: 20px;
  margin: 5px 0px 0px 0px;
  width: 20px;
}

.c0:hover .c2 {
  background-color: #A7A7A7;
}

.c3 {
  box-sizing: border-box;
  color: #242424;
  cursor: pointer;
  display: inline-block;
  float: left;
  height: 30px;
  font-family: "Open Sans",sans-serif;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  font-stretch: normal;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  margin: 0px;
  padding: 4px 0px 0px 5px;
  text-align: left;
  width: auto;
  white-space: nowrap;
}

<div
  className="c0 c1"
  disabled={false}
  onClick={[Function]}
>
  <div
    checked={false}
    className="c2"
  />
  <label
    checked={false}
    className="c3"
  />
</div>
`;

exports[`[SNAPSHOT][Componentes][Common][CheckBox] - Checkbox normal. Debe pintar el componente con los siguientes modificadores: (Checked: false). 2`] = `
.c1 {
  box-sizing: border-box;
  display: inline-block;
  height: 30px;
  margin: 0px;
  padding: 0px;
  width: auto;
}

.c0:hover .c4 {
  background-color: #A7A7A7;
}

.c2 {
  background-color: #1476FB;
  border: none;
  border-radius: 5px;
  box-sizing: border-box;
  cursor: pointer;
  display: inline-block;
  float: left;
  height: 20px;
  margin: 5px 0px 0px 0px;
  width: 20px;
}

.c0:hover .c2 {
  background-color: #11AAFF;
}

.c2:after {
  border: solid white;
  border-width: 0px 3px 3px 0px;
  content: "";
  display: block;
  height: 10px;
  margin: 2px 0px 0px 6px;
  width: 5px;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
}

.c3 {
  box-sizing: border-box;
  color: #242424;
  cursor: pointer;
  display: inline-block;
  float: left;
  height: 30px;
  font-family: "Open Sans",sans-serif;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  font-stretch: normal;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  margin: 0px;
  padding: 4px 0px 0px 5px;
  text-align: left;
  width: auto;
  white-space: nowrap;
}

<div
  className="c0 c1"
  disabled={false}
  onClick={[Function]}
>
  <div
    checked={true}
    className="c2"
  />
  <label
    checked={true}
    className="c3"
  />
</div>
`;
